typing/rec_check.ml

```diff
@@ -193,15 +193,26 @@ let array_element_kind env ty =
             when List.for_all (fun c -> c.Types.cd_args = Types.Cstr_tuple [])
                 cstrs ->
               `Pintarray
-          | {type_kind = _} ->
+          | {type_kind = Type_variant _ | Type_open | Type_record _} ->
               `Paddrarray
         with Not_found ->
           (* This can happen due to e.g. missing -I options,
               causing some .cmi files to be unavailable.
               Maybe we should emit a warning. *)
           `Pgenarray
       end
-  | _ ->
+  | Tlink _ | Tsubst _ ->
+      (* [Typeopt.scrape_ty] starts with [Ctype.expand_head_opt],
+         so these cases are impossible *)
+      assert false
```

Comment:
I agree that `Tlink` can't happen, because `repr` removes links.

I'm not sure your explanation holds for `Tsubst` though: `try_expand_head_opt` will raise `Cannot_expand` on `Tsubst`s, which will be caught by `expand_head_opt` which will then simply return its input (well, after calling `repr` on it).

Note: I'm not saying `Tsubst` can happen at this particular place, just that your explanation doesn't convince me.

Reply:
> and some of the things I wrote there are wishful thinking

Please, don't (ever) do that again.

> I agree that it's not clear what happens with `Tsubst` here, what motivated me to include it in the `assert false` is that other functions of `ctype.ml` failed on it in the same way.

Well, I guess that's because `Tsubst` can only appear in very few places, and are always (afaict) cleaned up right away.
It seems pretty obvious that your function is never called in these periods of time where `Tsubst` exists, so `assert false` is indeed fine.

I'd suggest either removing your comment, or splitting the cases and given a proper explanation to both.